common/lockmanager_test.go

```diff
@@ -0,0 +1,318 @@
+package common
+
+import (
+    "sync"
+    "testing"
+    "time"
+
+    "github.com/edulinq/autograder/config"
+)
+
+func TestLockBase(test *testing.T) {
+    key1 := "testkey1";
+
+    Lock(key1);
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+}
+
+func TestUnlockingAnUnlockedLock(test *testing.T) {
+    key1 := "testkey1";
+
+    Lock(key1);
+
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+
+    err = Unlock(key1);
+    if (err == nil) {
+        test.Fatalf("Failed to return an error when unlocking a lock that is already unlocked.");
+    }
+}
+
+func TestUnlockingKeyThatDoesntExist(test *testing.T) {
+    doesNotExistKey := "dne";
+
+    err := Unlock(doesNotExistKey);
+    if (err == nil) {
+        test.Fatalf("Failed to return an error when unlocking a key that does not exist.");
+    }
+}
+
+func TestLockUnlockDifferentKeysOneThread(test *testing.T) {
+    key1 := "testkey1";
+    key2 := "testkey2";
+    var verifyWaitGroup sync.WaitGroup;
+    var threadWaitGroup sync.WaitGroup;
+
+    // Lock key1 for the first time.
+    Lock(key1);
+
+    threadWaitGroup.Add(1);
+    verifyWaitGroup.Add(1);
+
+    // This thread should lock key2 for the first time while key1 is already locked.
+    go func() {
+        defer threadWaitGroup.Done();
+        Lock(key2);
+        verifyWaitGroup.Done();
+        err := Unlock(key2);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock.");
+        }
+    }()
+
+    // Make sure the thread locked key2 before the main thread unlocks key1.
+    verifyWaitGroup.Wait();
+
+    // Unlock key1.
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock.");
+    }
+
+    // Wait for the thread to finish unlocking key2.
+    threadWaitGroup.Wait();
+}
+
+func TestLockUnlockDifferentKeysTwoThreads(test *testing.T) {
+    key1 := "testkey1";
+    key2 := "testkey2";
+    var threadsWaitGroup sync.WaitGroup;
+    var funcWaitGroup sync.WaitGroup
+
+    threadsWaitGroup.Add(2);
+    funcWaitGroup.Add(2);
+
+    // First thread locks and unlocks key1.
+    go func() {
+        defer funcWaitGroup.Done();
+        Lock(key1);
+        threadsWaitGroup.Done();
+        err := Unlock(key1);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock.")
+        }
+    }()
+
+    // Second thread locks and unlocks key2.
+    go func() {
+        defer funcWaitGroup.Done();
+        Lock(key2);
+        threadsWaitGroup.Done();
+        err := Unlock(key2);
+        if (err != nil) {
+            test.Fatalf("Failed to Unlock")
+        }
+    }()
+    
+    // Wait until both threads have finished locking their keys so they 
+    // can unlock at the same time.
+    threadsWaitGroup.Wait();
```

Comment:
The threads are not waiting.
WaitGroup.Done() does not block.
https://pkg.go.dev/sync#WaitGroup.Done

Reply:
That comment is meant for the threadsWaitGroup.Wait() line, so I'm not too sure what you mean by this.

Reply:
This comment is in the correct spot.

WaitGroup.Done() does not block, so your other threads that called `threadsWaitGroup.Done()` are not synchronized.
Your comments indicates that you want those threads waiting at the same point, but they cannot be because Done() does not block.